File: frontend/src/lib/api/codebook.ts
import api from './client'
import type { MagnitudeScale } from '../magnitude'

// ── Types ────────────────────────────────────────────────────────────────────

export interface CodebookCodeNode {
  id: number
  numeric_id: number
  name: string
  description: string | null
  color: string | null
  is_active: boolean
  is_universal: boolean
  segment_count: number
  source_count: number
  excerpt_count: number
  category_id: number | null
  // #501: typed source identities ("conv:1", "col:13", "doc:2") — lets the
  // peek multi-select UNION sources instead of double-counting shared ones.
  source_keys?: string[]
  /** #869 (b): the code's declared rating scale, so the merge dialog can say
   * before a merge when two codes' scales differ. Null when it has none. */
  magnitude_scale?: MagnitudeScale | null
}

export interface CodebookCategoryNode {
  id: number
  name: string
  color: string | null
  display_order: number
  parent_id: number | null
  depth: number
  created_at: string | null
  code_count: number
  total_code_count: number
  total_segments: number
  total_sources: number
  children: CodebookCategoryNode[]
  codes: CodebookCodeNode[]
}

export interface CodebookTreeResponse {
  universal_codes: CodebookCodeNode[]
  tree: CodebookCategoryNode[]
  uncategorized_codes: CodebookCodeNode[]
}

// ── API ──────────────────────────────────────────────────────────────────────

export const codebookApi = {
  tree: (pid: number, params?: {
    conversation_ids?: string
    text_column_ids?: string
    exclude_facilitator?: boolean
    include_inactive?: boolean
    min_segments?: number
    max_segments?: number
  }) => api.get<CodebookTreeResponse>(`/projects/${pid}/codebook/tree`, { params })
    .then(r => r.data),
}

File: backend/app/schemas/codebook.py
from __future__ import annotations
from datetime import datetime
from .common import UTCTimestamp
from pydantic import BaseModel


class CodebookCodeNode(BaseModel):
    id: int
    numeric_id: int
    name: str
    description: str | None = None
    color: str | None = None
    is_active: bool
    is_universal: bool
    segment_count: int = 0
    source_count: int = 0
    excerpt_count: int = 0
    category_id: int | None = None
    # #501: typed source identities ("conv:1", "col:13", "doc:2") so the peek
    # panel's multi-select can UNION sources across codes — Σ of per-code
    # source_counts double-counts shared sources (it exceeded the project's
    # source universe on the audit corpus).
    source_keys: list[str] = []


class CodebookCategoryNode(BaseModel):
    id: int
    name: str
    color: str | None = None
    display_order: int
    parent_id: int | None = None
    depth: int = 0
    created_at: UTCTimestamp | None = None
    code_count: int = 0
    total_code_count: int = 0
    total_segments: int = 0
    total_sources: int = 0
    children: list[CodebookCategoryNode] = []
    codes: list[CodebookCodeNode] = []


class CodebookTreeResponse(BaseModel):
    universal_codes: list[CodebookCodeNode] = []
    tree: list[CodebookCategoryNode] = []
    uncategorized_codes: list[CodebookCodeNode] = []
